File: api_server/models/post.py
from . import database
from GFMarkDown import markdown
from datetime import datetime
from .relations import post_catalog, post_tag
from sqlalchemy.sql import func
from .user import User

db = database.db


class Post(db.Model):
    post_id = db.Column(db.Integer, primary_key=True)
    display_id = db.Column(db.Text, unique=True)
    title = db.Column(db.String)
    subtitle = db.Column(db.Text, default='')
    content = db.Column(db.Text, default='')
    md_content = db.Column(db.Text, default='')
    create_time = db.Column(db.DateTime, server_default=func.now())
    last_edit_time = db.Column(db.DateTime, server_default=func.now(), onupdate=func.now())
    owner = db.Column(db.Integer, db.ForeignKey('user.user_id'))
    cover = db.Column(db.Text)
    secret = db.Column(db.Integer, nullable=False, default=0)
    hidden = db.Column(db.Boolean, nullable=False, default=False)
    dark_title = db.Column(db.Boolean, nullable=False, default=False)
    enable_comment = db.Column(db.Boolean, nullable=False, default=True)
    catalogs = db.relationship(
        'Catalog',
        secondary=post_catalog,
        backref='post'
    )
    author = db.relationship(
        'User',
        uselist=False,
        backref='post'
    )
    tag = db.relationship(
        'Tag',
        secondary=post_tag,
        backref='post'
    )

    comments = db.relationship(
        'Comment',
        backref='post',
        cascade='all,delete'
    )

    def __init__(self, **kwargs):
        self.title = kwargs['title']
        self.display_id = kwargs.get('display_id', self.title) or None
        self.subtitle = kwargs.get('subtitle', '')
        content = kwargs.get('content')
        md_content = kwargs.get('md_content')
        if content is None and md_content:
            content = markdown(md_content)
        self.content = content
        self.md_content = md_content
        if 'owner' in kwargs:
            self.owner = kwargs['owner']
        if 'author' in kwargs:
            self.author = kwargs['author']
        self.cover = kwargs.get('cover', '')
        self.secret = kwargs.get('secret', 0)
        self.dark_title = kwargs.get('dark_title', False)
        self.enable_comment = kwargs.get('enable_comment', True)
        self.tag = kwargs.get('tags', [])
        self.catalogs = kwargs.get('catalogs', [])

    @property
    def dict(self):
        return {
            'id': self.post_id,
            'display_id': self.display_id,
            'title': self.title,
            'subtitle': self.subtitle,
            'content': self.content,
            'md_content': self.md_content,
            'created': self.create_time,
            'last_edit': self.last_edit_time,
            'date': self.create_time.strftime('%b %d, %Y'),
            'img': self.cover,
            'dark_title': self.dark_title,
            'enable_comment': self.enable_comment,
            'author': self.author and self.author.dict,
            'secret': self.secret,
            'hidden': self.hidden,
            'tags': [x.tag_name for x in self.tag],
            'catalogs': [x.catalog_name for x in self.catalogs]
        }

    @property
    def brief_dict(self):
        return {
            'id': self.post_id,
            'display_id': self.display_id,
            'title': self.title,
            'subtitle': self.subtitle,
            'created': self.create_time,
            'date': self.create_time.strftime('%b %d, %Y'),
            'last_edit': self.last_edit_time,
            'img': self.cover,
            'dark_title': self.dark_title,
            'enable_comment': self.enable_comment,
            'author': self.author and self.author.dict,
            'secret': self.secret,
            'hidden': self.hidden,
            'tags': [x.tag_name for x in self.tag],
            'catalogs': [x.catalog_name for x in self.catalogs]
        }

    modify_forbidden_attrs = {
        'created', 'last_edit', 'author', 'post_id', 'id'
    }

File: RoseliaCore/Database/Models/Post.fs
namespace rec RoseliaBlog.RoseliaCore.Database.Models

open System
open System.Collections.Generic
open System.ComponentModel
open System.ComponentModel.DataAnnotations
open System.ComponentModel.DataAnnotations.Schema
open System.Text.Json.Serialization
open System.Text.RegularExpressions
open Microsoft.EntityFrameworkCore

[<CLIMutable>]
[<Table("post")>]
[<Index(propertyName="DisplayId", IsUnique = true)>]
type Post = {
    [<Key>]
    [<Column("post_id")>]
    mutable PostId: int
    
    [<Column("display_id")>]
    mutable DisplayId: string
    
    [<Column("title")>]
    mutable Title: string
    
    [<Column("subtitle")>]
    [<DefaultValue("")>]
    mutable Subtitle: string
    
    [<Column("content")>]
    mutable Content: string
    
    [<Column("md_content")>]
    mutable MarkdownContent: string
    
    [<Column("create_time")>]
    [<DatabaseGenerated(DatabaseGeneratedOption.Identity)>]
    [<DataType(DataType.DateTime)>]
    CreatedTime: DateTime
    
    [<Column("last_edit_time")>]
    [<DatabaseGenerated(DatabaseGeneratedOption.Computed)>]
    [<DataType(DataType.DateTime)>]
    LastEditedTime: DateTime
    
    [<Column("owner")>]
    Owner: int
    
    [<ForeignKey("Owner")>]
    Author: User
    
    [<Column("cover")>]
    mutable Cover: string
    
    [<Column("secret")>]
    [<DefaultValue(0)>]
    mutable Secret: int
    
    [<Column("hidden")>]
    [<DefaultValue(false)>]
    mutable Hidden: bool
    
    [<Column("dark_title")>]
    [<DefaultValue(false)>]
    mutable IsTitleDark: bool
    
    [<Column("enable_comment")>]
    [<DefaultValue(true)>]
    [<JsonPropertyName("enable_comment")>]
    mutable IsCommentEnabled: bool
    
    Tags: Tag ICollection
    Catalogs: Catalog ICollection
}

[<CLIMutable>]
[<Table("catalog")>]
type Catalog = {
    [<Key; Column("catalog_id")>]
    CatalogId: int
    
    [<Column("catalog_eternal_link")>]
    CatalogEternalLink: string
    
    [<Column("catalog_name")>]
    CatalogName: string
    
    Posts: Post ICollection
}

[<CLIMutable>]
[<Table("tag")>]
[<Index(propertyName="TagName", IsUnique = true)>]
type Tag = {
    [<Key; Column("tag_id")>]
    TagId: int
    
    [<Column("tag_name"); MaxLength(64)>]
    TagName: string
    
    Posts: Post ICollection
}

module Tag =
    let GetDisplayedName name =
        Regex.Replace(name, @"\s+", "-")
